lib/EnzymeTestUtils/src/to_vec.jl

```diff
@@ -0,0 +1,66 @@
+# alternative to FiniteDifferences.to_vec to use Enzyme's semantics for arrays instead of
+# ChainRules': Enzyme treats tangents of AbstractArrays the same as tangents of any other
+# struct (i.e. with a container of the same type as the original), while ChainRules
+# represents the tangent with an array of some type that is tangent to the subspace defined
+# by the original array type.
+
+# base case: we've unwrapped to a number, so we break the recursion
+to_vec(x::AbstractFloat) = ([x], only)
+
+# types: they and their fields aren't differentiable, so we bypass them
+to_vec(x::Type) = (Float32[], _ -> x)
+
+# below code is adapted from https://github.com/JuliaDiff/FiniteDifferences.jl/blob/99ad77f05bdf6c023b249025dbb8edc746d52b4f/src/to_vec.jl
+# MIT Expat License
+# Copyright (c) 2018 Invenia Technical Computing
+
+# structs: recursively call to_vec on each field
+function to_vec(x::T) where {T}
```

Comment:
Does this need to be made with a map of seen variables (e.g. if you have tuple{Vector, Vector} and x[0] === x[1])?

Reply:
I suppose since we only take type's here and not variables, that doesn't apply here

Reply:
No, I don't think it should matter whether e.g. a tuple contains the same array twice. But we should test for that. If that is a problem, it will put a major restriction on testable functions. We could at least detect that and throw an informative error pretty easily I think.

Reply:
(a limitation we cannot let go of is that the primal and the shadow share the same memory pattern. e.g. if `x` is an array, Enzyme is fine with a user providing `Duplicated((x, x), (dx, copy(dx)))` to `autodiff`, but for testing we would need to require `Duplicated((x, x), (dx, dx))`.)

Reply:
Is there a reason why zero_tangent can't just call make_zero? The make_zero logic is intentionally general (and if needed user extensible like deepcopy), to deal with all the sorts of crazy struct/etc things.

Reply:
I see what you mean by the to_vec unifying simplicity, but we need make_zero to be potentially "easily" user-extensible for custom types and I don't want to make that burden harder unless necessary (even though it is compelling here to simplify code).

Reply:
> Is there a reason why zero_tangent can't just call make_zero?

The main reason is we need aliased arrays to in the primal to also have aliased shadows for our code to do the right thing. `zero_tangent` doesn't currently handle that but could. `make_zero` doesn't either (because it relies on `deepcopy_internal`, which also doesn't).

I guess it depends on whether we think those user-extensions of `make_zero` would ever be for structs one might want to pass to EnzymeTestUtils. If so, the user would also need to similarly extend `zero_tangent`.